---
title: "Processing VCF output from base editing data"
author: "Angelo Limeta"
date: "14/4/2020"
output:
  html_document:
    df_print: kable
    toc: true
    toc_float: true
---

## Load libraries
```{r}
library(tidyverse)
select = dplyr::select
mutate = dplyr::mutate
rename = dplyr::rename
```

## Load Data

Load .vcf file and sample metadata
```{r}
df_vcf <- read_delim("../data/mpileup.tsv", 
    "\t", escape_double = FALSE, trim_ws = TRUE)
head(df_vcf)

df_meta <- read_delim("../data/meta.txt", 
    "\t", escape_double = FALSE, trim_ws = TRUE)
head(df_meta)
```

## Process vcf file

The .vcf file contains allele counts for each genomic position across samples. These are denoted as AD in the FORMAT field, and correspond to the last entry in each sample.
```{r}
df_vcf$FORMAT[1]
```

For example, let's take a look at the alleles detected in sample "P17460_1032_S32_L001" at genomic position 72.
```{r}
cat("Reference allele: ")
df_vcf %>% 
  filter(POS == 72) %>% 
  select(REF) %>% 
  as.character() %>% 
  cat()

cat("\nAlternative alleles: ")
df_vcf %>% 
  filter(POS == 72) %>% 
  select(ALT) %>% 
  as.character() %>% 
  cat()

```

Counts for each allele in this particular sample can be accessed as by filtering out the AD entry:
```{r}
allele_counts = df_vcf %>% 
  filter(POS == 72) %>% 
  select(P17460_1032_S32_L001) %>% 
  as.character() %>%
  strsplit(split = ":")
allele_counts = allele_counts[[1]][6] #AD is the sixth entry
cat("Allele counts:\n")
cat(allele_counts)
```
__These correspond to:__
* 1 count for the "G" allele, i.e. the reference allele
* 6 counts for the "A" allele
* 0 counts for the "*" allele

Let's convert the matrix into long format for easier downstream processing
```{r}
df_long = df_vcf %>%
  pivot_longer(-c(CHROM,POS,ID,REF,ALT,QUAL,FILTER,INFO,FORMAT),
               names_to = "SAMPLE",
               values_to = "FREQ")
head(df_long)
```
Extract only the AD field and combine all alleles into one column
```{r}
df_long = df_long %>% 
  mutate(FREQ = gsub(".*:","",FREQ)) %>% 
  mutate(ALLELES = paste(REF,ALT,sep = ",")) %>% 
  mutate(ALLELES = gsub("<\\*>","-",ALLELES))

df_long %>% 
  select(ALLELES,FREQ) %>% 
  head()
```
Let's write a function that converts this so that each allele gets its own row
```{r}
tidyAlleles = function(df)
{
  # Calculate how many rows are required
  new_no_rows = 0
  for (i in 1:nrow(df)) {
    no_alleles_at_pos = length(strsplit(df$FREQ[i], split = ",")[[1]])
    new_no_rows = new_no_rows + no_alleles_at_pos
  }
  
  # Initialize empty (long) tibble
  df_output = df[1, ] %>%
    add_row(SAMPLE = rep("test", new_no_rows - 1)) %>%
    mutate(ALLELE = NA,
           FREQ_ALLELE = NA,
           ALLELE_TYPE = NA)
  
  # Fill tibble with values
  # Indexing variable for keeping track of the row number in the tidy tibble
  index = 1
  for (i in 1:nrow(df)) {
    # Obtain list of alleles and frequencies for each position in each sample
    alleles_at_pos = strsplit(df$ALLELES[i], split = ",")[[1]]
    freq_at_pos = strsplit(df$FREQ[i], split = ",")[[1]]
    for (j in 1:length(alleles_at_pos)) {
      # fill rows from the original tibble
      df_output[index, 1:ncol(df)] = df[i, 1:ncol(df)]
      # Add allele and frequency
      df_output$ALLELE[index] = alleles_at_pos[j]
      df_output$FREQ_ALLELE[index] = freq_at_pos[j]
      # Add allele type (REF,SNP,INDEL)
      if (nchar(paste0(alleles_at_pos[1], alleles_at_pos[j])) > 2) {
        df_output$ALLELE_TYPE[index] = "INDEL"
      } else if (alleles_at_pos[1] == alleles_at_pos[j]){
        df_output$ALLELE_TYPE[index] = "REF"
      } else {
        df_output$ALLELE_TYPE[index] = "SNP"
      }
      index = index + 1
    }
  }
  
  return(df_output)
}

# df_temp = tidyAlleles(df_long)
# Actually, tidyverse has a much faster solution
df_tidy = separate_rows(df_long,FREQ,ALLELES,sep = ",") %>%
  rename(ALLELE = ALLELES) %>%
  relocate(FREQ, .after = last_col()) %>% 
  relocate(SAMPLE, .before = CHROM) %>%
  mutate(MUT_TYPE = case_when(
    (nchar(REF) != nchar(ALLELE)) ~ "INDEL",
    ((nchar(REF) == nchar(ALLELE)) & (REF == ALLELE)) ~ "REF",
    ((nchar(REF) == 1 | nchar(ALLELE) == 1) & (REF != ALLELE)) ~ "SNP"
  )) %>% 
  mutate(FREQ = as.integer(FREQ),
         POS = as.integer(POS))

df_tidy %>% 
  select(SAMPLE, POS,ALLELE,FREQ,MUT_TYPE) %>% 
  head()
```
Let's add read depth information at each position
```{r}
df_tidy = df_tidy %>% 
  group_by(SAMPLE,POS) %>% 
  mutate(READ_DEPTH = sum(FREQ)) %>% 
  ungroup()
```


## Process metadata

Let's begin by extracting target gene, time, and condition.
```{r}
df_meta = df_meta %>% 
  separate(`User ID`, c("TARGET","TIME","CONDITION"),sep = "_",remove = FALSE) %>% 
  rename(USER_ID=`User ID`)

head(df_meta)
```
Let's also extract editor and gRNA information
```{r}
df_meta = df_meta %>% 
  mutate(GUIDE = str_extract(CONDITION, "gR\\d$"),
         EDITOR = str_extract(CONDITION, "dCas\\w{3}"),
         REPLICATE = str_extract(CONDITION, "\\d$")) %>% 
  mutate(EDITOR = ifelse(is.na(EDITOR),"WT",EDITOR))

head(df_meta)
```
## Merge data

Let's shorten the SAMPLE ID in order to match the NGI ID in the metadata
```{r}
df_tidy = df_tidy %>% 
  mutate(SAMPLE = substr(SAMPLE,1,11))

head(df_tidy)
```
Merge both data sets
```{r}
df_full = df_tidy %>% 
  inner_join(df_meta,by = c("SAMPLE" = "NGI ID")) %>%
  select(-c(SAMPLE)) %>% 
  rename(SAMPLE = USER_ID,
         VCF_INFO = INFO) %>% 
  relocate(SAMPLE, .before = CHROM) %>% 
  select(-c(CHROM))
head(df_full)
```

## Save

```{r}
saveRDS(df_full, file = "../data/processed_data/christos_new.rds")
```

