---
title: "Processing VCF output from base editing data"
author: "Angelo Limeta"
date: "14/4/2020"
output:
  html_document:
    df_print: kable
    toc: true
    toc_float: true
---

# Load libraries
```{r}
library(tidyverse)
library(cowplot)
library(ggpubr)
library(ggsci)
library(hrbrthemes)
library(ggrepel)
library(ggExtra)
select = dplyr::select
mutate = dplyr::mutate
```

# Load Data

Load pre-processed vcf data and metadata
```{r}
df_seq <- readRDS(file = "../data/processed_data/christos_new.rds")
head(df_seq)
```

## Plot non-normalized data

Let's begin by plotting read depth across the sequence per sample
```{r}
df_seq %>%
  group_by(SAMPLE) %>% 
  summarise(median_read_depth = median(READ_DEPTH)) %>% 
  ungroup() %>% 
  head()

df_seq %>% 
  ggplot(aes(x=POS,y=READ_DEPTH,group=SAMPLE, color=SAMPLE)) +
  geom_point() +
  theme(legend.position = "none")
```

Let's look at the amount of mutations per sample, and compare this to read depth
```{r}
df_seq %>%
  group_by(SAMPLE) %>% 
  summarise(median_read_depth = median(READ_DEPTH),
            median_mutation_freq = median(FREQ)) %>% 
  ungroup()
```

# Data pre-processing

## Adjust for read depth and end regions

Remove low coverage regions
```{r}
df_seq = df_seq %>% 
  filter(READ_DEPTH > 1000)

df_seq %>% head()
```

Remove end regions
```{r}
df_seq = df_seq %>% 
  filter(POS > 80) %>% 
  filter(POS < 210)

df_seq %>% head()
```

Normalize by sequencing depth
```{r}
df_seq = df_seq %>% 
  group_by(SAMPLE,POS) %>% 
  mutate(NORM_FREQ = FREQ/(READ_DEPTH),
         LOG_NORM_FREQ = log2(FREQ/READ_DEPTH)) %>% 
  ungroup()

df_seq %>% head()
```

Let's calculate the frequency of all alternative alleles at each position
```{r}
df_seq = df_seq %>%
  group_by(SAMPLE, POS, ) %>%
  mutate(MUT_FREQ = ifelse(MUT_TYPE == "SNP",
                           sum(NORM_FREQ[MUT_TYPE == "SNP"]),
                           NA)) %>%
  ungroup()
```


Let's now subtract the T0 alternative alleles frequency. Let's define this as the mutation enrichment of a given base.
```{r}
df_seq = df_seq %>%
  group_by(CONDITION, POS) %>%
  mutate(MUT_ENRICH = ifelse(MUT_TYPE == "SNP",
                                MUT_FREQ/MUT_FREQ[TIME == "T0"],
                                MUT_FREQ)) %>%
  mutate(MUT_ENRICH = ifelse(MUT_ENRICH < 0,
                               0,
                               MUT_ENRICH)) %>%
  ungroup() 
```



# Analysis of all mutations

## Plot mutations accross the gene sequence

Configure graphical parameters
```{r}
# y-axis limits
ylim_max = 10
# PAM
anno_rect_PAM = annotate("rect", xmin = 138, xmax = 140, ymin = -Inf, ymax = Inf, alpha=1, fill="black")
# gRNA binding site
anno_rect_gRNA = annotate("rect", xmin = 127, xmax = 147, ymin = -Inf, ymax = Inf, alpha=0.3, fill="black")
# +/-50bp editing window
anno_rect_baseEditWindow = annotate("rect", xmin = 77, xmax = 197, ymin = -Inf, ymax = Inf, alpha=0.1, fill="black")

```

Plot mutation enrichment for MAG editor (gRNA binding site in Grey)
```{r,fig.height=6,fig.width=6}

df_seq %>% 
  filter(grepl("MAG1gR",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "qual",palette = "Set1") +
  theme(legend.position = "none") +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  theme(aspect.ratio = 0.7) +
  ylim(0,ylim_max)

df_seq %>% 
  filter(grepl("MAG\\d$",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)

df_seq %>% 
  filter(grepl("WT",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)
```

Plot mutation enrichment for AID editor (gRNA binding site in Grey)
```{r,fig.height=6,fig.width=6}

df_seq %>% 
  filter(grepl("AIDgR",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "qual",palette = "Set1") +
  theme(legend.position = "none") +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  theme(aspect.ratio = 0.7) +
  ylim(0,ylim_max)

df_seq %>% 
  filter(grepl("AID\\d$",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)

df_seq %>% 
  filter(grepl("WT",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)
```

## Compare mutation enrichment with read_depth

Let's sum up the mutation enrichment across the gene in each sample and plot against the read depth
```{r}
df_cum_mut = df_seq %>% 
  filter(MUT_TYPE == "SNP") %>% 
  filter(TIME != "T0") %>%
  group_by(SAMPLE) %>% 
  summarise(CUM_MUT_ENRICH = sum(MUT_ENRICH),
            MEDIAN_READ_DEPTH = median(READ_DEPTH),
            EDITOR = EDITOR,
            TIME = TIME) %>%
  ungroup() %>% 
  distinct() %>% 
  filter(!is.na(CUM_MUT_ENRICH)) %>% 
  mutate(GRNA = ifelse(grepl("gR",SAMPLE),"Guide","No-guide"))
head(df_cum_mut)
```


```{r, fig.width= 10, fig.height=6}
# Color by editor
p_cum1 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = EDITOR)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "dCasMAG" = "dodgerblue2",
    "dCasAID" = "violetred",
    "WT" = "black"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("") +
  ylab("Cumulative mutation enrichment")

# Color by gRNA
p_cum2 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = GRNA)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "Guide" = "cyan4",
    "No-guide" = "darkorange1"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("Median read depth") +
  ylab("")

# Color by transfer
p_cum3 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = TIME)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "T1" = "gray55",
    "T5" = "black"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("") +
  ylab("")

# Add marginal distributions
p_cum1 = ggMarginal(p_cum1, type = "boxplot", size = 10,groupColour = TRUE)
p_cum2 = ggMarginal(p_cum2, type = "boxplot", size = 10,groupColour = TRUE)
p_cum3 = ggMarginal(p_cum3, type = "boxplot", size = 10,groupColour = TRUE)

ggarrange(p_cum1, p_cum2, p_cum3,
          ncol = 3)
```

Hmmm, seems like the presence of the gRNA didn't change much.
Let's try shrinking our data to only look at the editing window, i.e. the gRNA binding site.
```{r}
df_cum_mut = df_seq %>% 
  filter(MUT_TYPE == "SNP") %>% 
  filter(POS >= 127) %>% 
  filter(POS <= 147) %>% 
  filter(TIME != "T0") %>%
  group_by(SAMPLE) %>% 
  summarise(CUM_MUT_ENRICH = sum(MUT_ENRICH),
            MEDIAN_READ_DEPTH = median(READ_DEPTH),
            EDITOR = EDITOR,
            TIME = TIME) %>%
  ungroup() %>% 
  distinct() %>% 
  filter(!is.na(CUM_MUT_ENRICH)) %>% 
  mutate(GRNA = ifelse(grepl("gR",SAMPLE),"Guide","No-guide"))
head(df_cum_mut)
```

```{r, fig.width= 10, fig.height=6}
# Color by editor
p_cum1 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = EDITOR)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "dCasMAG" = "dodgerblue2",
    "dCasAID" = "violetred",
    "WT" = "black"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("") +
  ylab("Cumulative mutation enrichment")

# Color by gRNA
p_cum2 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = GRNA)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "Guide" = "cyan4",
    "No-guide" = "darkorange1"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("Median read depth") +
  ylab("")

# Color by transfer
p_cum3 = df_cum_mut %>%
  ggplot(aes(x = MEDIAN_READ_DEPTH, y = CUM_MUT_ENRICH, color = TIME)) +
  geom_point(size = 3) +
  scale_color_manual(values = c(
    "T1" = "gray55",
    "T5" = "black"
  )) +
  theme_bw() +
  theme(aspect.ratio = 1,
        legend.position = "bottom") +
  labs(size = "Transfer",color = "") +
  xlab("") +
  ylab("")

# Add marginal distributions
p_cum1 = ggMarginal(p_cum1, type = "boxplot", size = 10,groupColour = TRUE)
p_cum2 = ggMarginal(p_cum2, type = "boxplot", size = 10,groupColour = TRUE)
p_cum3 = ggMarginal(p_cum3, type = "boxplot", size = 10,groupColour = TRUE)

ggarrange(p_cum1, p_cum2, p_cum3,
          ncol = 3)
```

# Filter out and analyze C -> T edits

Since the base editor can only act on cytidine, we could try to filter out only these edits from our data
```{r}
df_seq_cyt = df_seq %>% 
  filter(REF == "C")
head(df_seq_cyt)
```
## Plot mutations accross the gene sequence

Configure graphical parameters
```{r}
# y-axis limits
ylim_max = 30
# PAM
anno_rect_PAM = annotate("rect", xmin = 138, xmax = 140, ymin = -Inf, ymax = Inf, alpha=1, fill="black")
# gRNA binding site
anno_rect_gRNA = annotate("rect", xmin = 127, xmax = 147, ymin = -Inf, ymax = Inf, alpha=0.3, fill="black")
# +/-50bp editing window
anno_rect_baseEditWindow = annotate("rect", xmin = 77, xmax = 197, ymin = -Inf, ymax = Inf, alpha=0.1, fill="black")

```

Plot mutation enrichment for MAG editor (gRNA binding site in Grey)
```{r,fig.height=6,fig.width=6}

df_seq_cyt %>% 
  filter(grepl("MAG1gR",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "qual",palette = "Set1") +
  theme(legend.position = "none") +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  theme(aspect.ratio = 0.7) +
  ylim(0,ylim_max)

df_seq_cyt %>% 
  filter(grepl("MAG\\d$",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)

df_seq_cyt %>% 
  filter(grepl("WT",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)
```

Plot mutation enrichment for AID editor (gRNA binding site in Grey)
```{r,fig.height=6,fig.width=6}

df_seq_cyt %>% 
  filter(grepl("AIDgR",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "qual",palette = "Set1") +
  theme(legend.position = "none") +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  theme(aspect.ratio = 0.7) +
  ylim(0,ylim_max)

df_seq_cyt %>% 
  filter(grepl("AID\\d$",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)

df_seq_cyt %>% 
  filter(grepl("WT",CONDITION)) %>% 
  filter(MUT_TYPE == "SNP") %>%
  ggplot(aes(x=POS,y=MUT_ENRICH, color = CONDITION)) +
  anno_rect_baseEditWindow +
  anno_rect_gRNA +
  anno_rect_PAM +
  geom_point() +
  facet_wrap(~ TIME + CONDITION,ncol = 3) +
  theme_bw() +
  scale_color_brewer(type = "div",palette = "Set1") +
  theme(legend.position = "none",
        aspect.ratio = 0.7) +
  ylab("Mutation enrichment") +
  xlab("Position on gene") +
  ylim(0,ylim_max)
```
